### PR TITLE
Add staging directive to rebase script

### PR DESCRIPTION
**Pull request by fzdarsky**
Extension of the rebase script to add the "staging" directive used by PR #804.

Signed-off-by: Frank A. Zdarsky [fzdarsky@redhat.com](mailto:fzdarsky@redhat.com)

[USHIFT-224](https://issues.redhat.com//browse/USHIFT-224)

### CODE REVIEW

**Comment by benluddy** [comment]
If we pick up a new indirect dependency on a module from staging, will someone need to remember to add an explicit require directive and "staging" replace directive? It might be simplest to force a `require k8s.io/foo v0.0.0` for all direct and indirect dependencies from staging, then generate a `replace k8s.io/foo v0.0.0 => <replacement from openshift/kubernetes>` for each, instead of making the existing replace directives the source of truth.

File: scripts/rebase.sh [hunk 1]
```diff
@@ -427,7 +443,7 @@ rebase_to() {
         title "## Updating vendor directory"
         go mod vendor
         title "## Patching vendor directory"
-        git apply scripts/rebase_patches/*
+        git apply scripts/rebase_patches/*.patch 2>/dev/null || true
```
Comment (dhellmann) [inline scripts/rebase.sh:446]: Why hide the errors from this step?

**Comment by dhellmann** [comment]
> If we pick up a new indirect dependency on a module from staging, will someone need to remember to add an explicit require directive and "staging" replace directive? It might be simplest to force a `require k8s.io/foo v0.0.0` for all direct and indirect dependencies from staging, then generate a `replace k8s.io/foo v0.0.0 => <replacement from openshift/kubernetes>` for each, instead of making the existing replace directives the source of truth.

The simplicity of forcing the replacements appeals to me. Are there any downsides?

**Comment by benluddy** [comment]
> Are there any downsides?

I don't think so. It's effectively the same approach, except the source of truth becomes the list of direct and indirect requires instead of the existing replacements.

File: scripts/rebase.sh (hunk 1 above)
Comment (mangelajo) [reply scripts/rebase.sh:446]: Right, I'd keep those,  otherwise it can be hard to trace

Comment (fzdarsky) [reply scripts/rebase.sh:446]: Soon having patches to apply will be the exception and I was bothered by CI highlighting the expected. But I agree and will instead apply conditionally on whether patches exist.

Comment (fzdarsky) [reply scripts/rebase.sh]: done

**Comment by dhellmann** [comment]
/override ci/prow/e2e-rpm-install

This PR doesn't change anything that runs in that job.

File: scripts/rebase.sh [hunk 2]
```diff
@@ -92,6 +95,34 @@ download_release() {
...
+
+# Updates a replace directive using an embedded component's commit.
+# Caches component pseudoversions for faster processing.
+declare -A pseudoversions
+replace_using_component_commit() {
+    local modulepath=$1
+    local new_modulepath=$2
+    local component=$3
+
+    if [ "${pseudoversions[${component}]+foo}" ]; then
```
Comment (dhellmann) [inline scripts/rebase.sh:113]: This test doesn't look right, there's no comparison. Should it have a `-z` or `-n` flag?

Comment (fzdarsky) [reply scripts/rebase.sh]: I agree this shortcut isn't very clear. The empty string basically tests as true with single '['. Updated this to be more [idiomatic](https://mywiki.wooledge.org/BashFAQ/046).

**Comment by fzdarsky** [comment]
Updated the commit to ensure Require- and ReplaceDirectives for all o/k staging repos to fix a few corner cases. Squashed with the previous attempt for easier readbility.

**Comment by fzdarsky** [comment]
Also used the oppy to fix an issue where the rebase script overwrites the MicroShift-specific ovn-kubernetes container image with the one from the OpenShift release image. Once both variants get merged and made available in the release image, this workaround can be removed again.

File: scripts/rebase.sh (hunk 2 above)
Comment (dhellmann) [reply scripts/rebase.sh]: Oh, wow, I'd never seen that before in bash. I guess it's just like testing an empty string in python.

**Comment by dhellmann** [comment]
I checked this out and tried with 4.10.26 and most everything worked except one of the steps close to the end

```
$ ./scripts/rebase.sh manifests
Extracting timestamp
Rebasing manifests
yq: -i/--in-place can only be used with -y/-Y
Script exited with error.
```

I don't know if that's caused by a change in this PR, a change in the release image, or a change in yq.

**Comment by fzdarsky** [comment]
> I'm on RHEL
> 
> ```
> $ yq --version
> yq 2.12.2
> ```

I'm on RHEL8, too, but my `yq` version is vastly different (it may even be we're using different tools):
```
$ yq --version
yq (https://github.com/mikefarah/yq/) version 4.26.1
```
I've installed mine as in https://github.com/openshift/release/blob/master/ci-operator/step-registry/openshift/microshift/rebase/openshift-microshift-rebase-commands.sh#L18-L27.

**Comment by dhellmann** [comment]
> > I'm on RHEL
> > ```
> > $ yq --version
> > yq 2.12.2
> > ```
> 
> I'm on RHEL8, too, but my `yq` version is vastly different (it may even be we're using different tools):
> 
> ```
> $ yq --version
> yq (https://github.com/mikefarah/yq/) version 4.26.1
> ```
> 
> I've installed mine as in https://github.com/openshift/release/blob/master/ci-operator/step-registry/openshift/microshift/rebase/openshift-microshift-rebase-commands.sh#L18-L27.

That explains it. I'm sure I just installed mine via yum. I'll replace it with this version and try again.

**Comment by dhellmann** [comment]
Replacing the python version of yq I had with the one mentioned above made this work.

/lgtm

**Pull request merged**